torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -2348,7 +2348,18 @@ void ProcessGroupNCCL::watchdogHandler() {
         // watchdog would unstage the stashed tensors when detecting completion
         // of the collective, to prevent ProcessGroupNCCL from holding reference
         // to those tensors forever.
-        work.unstashTensors();
+        // work.unstashTensors();
+        // Update: it seems directly unstashing from watchdog thread would cause
+        // some rare problems. We thus move the unstashing to main thread,
+        // triggered by a next user call, see `workEnqueue`. But `work` is going
+        // to be destructed, so we transfer the work's shelf to a shelves
+        // structure owned by the PG.
+        if (!work.stashed_for_allocator_safety_->empty()) {
+          std::lock_guard<std::mutex> lock(shelvesMutex_);
+          // We are just pushing back a shared_ptr here, so the cost should be
+          // minimal
+          shelvesToUnstash_.push_back(work.stashed_for_allocator_safety_);
```

Comment:
do we need to clear the work vector?

Reply:
`shelvesToUnstash_` is cleared later in `workEnqueue()` (so are the individual shelves in it).